jax/_src/extend/ffi.py

```diff
@@ -237,12 +238,26 @@ def ffi_call(
   else:
     multiple_results = False
     result_avals = _result_avals((result_shape_dtypes,))
+
+  # Because we support passing numpy arrays (which are not hashable) as
+  # attributes, we need to handle that here. For any non-hashable kwargs, we
+  # wrap them in a hashable object that uses the object's id as the hash. Then,
+  # in the lowering, we unwrap the object and use the original object.
+  hashable_kwargs = {}
+  for k, v in kwargs.items():
+    try:
+      hash(v)
+    except TypeError:
+      hashable_kwargs[k] = _HashableByObjectId(v)
+    else:
+      hashable_kwargs[k] = v
```

Comment:
Note: we could imagine using [`ad_checkpoint.WrapHashably`](https://github.com/jax-ml/jax/blob/20122ff7ce1eb44d266621045821b11e554dafcc/jax/_src/ad_checkpoint.py#L357-L378) here instead of [`api_util._HashableByObjectId`](https://github.com/jax-ml/jax/blob/20122ff7ce1eb44d266621045821b11e554dafcc/jax/_src/api_util.py#L725-L732) since they're roughly equivalent, but I wanted to avoid wrapping arguments that didn't need to be wrapped to keep the Jaxpr clean.